py/server/deephaven/barrage.py

```diff
@@ -0,0 +1,222 @@
+        and when BarrageSession.close() is called, it will shut down the channel as well as close the j_barrage_session..
+
+        when BarrageSession is initialized directly and j_managed_channel is None, when BarrageSession.close() is called,
+        it will only close the j_barrage_session, it is the responsibility of the user to shut down the associated
+        channel if needed.
+        """
+        self.j_barrage_session = j_barrage_session
+        self.j_session = j_barrage_session.session()
+        self.j_managed_channel = j_managed_channel
+
+    def __enter__(self) -> BarrageSession:
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+        self.close()
+
+    def close(self) -> None:
+        """ Closes the barrage session.
+
+        If the BarrageSession is initialized with a managed channel, the channel will be shut down as well.
+        """
+        try:
+            self.j_barrage_session.close()
+            if self.j_managed_channel:
+                self.j_managed_channel.shutdownNow()
+                self.j_managed_channel.awaitTermination(10, _JTimeUnit.SECONDS)
+        except Exception as e:
+            raise DHError(e, "failed to close the barrage session.") from e
+
+    def subscribe(self, ticket: bytes) -> Table:
+        """ Subscribes to a published remote table with the given ticket.
+
+        Note, if the remote table is closed or its owner session is closed, the ticket becomes invalid. If the same
+        ticket is subscribed to multiple times, multiple subscriptions will be created.
+
+        Args:
+            ticket (bytes): the bytes of the ticket
+
+        Returns:
+            a Table
+
+        Raises:
+            DHError
+        """
+        try:
+            j_table_handle = self.j_session.of(_JTableSpec.ticket(ticket))
+            j_barrage_subscription = self.j_barrage_session.subscribe(j_table_handle,
+                                                                      _JBarrageTableResolver.SUB_OPTIONS)
+            return Table(j_barrage_subscription.entireTable().get())
+        except Exception as e:
+            raise DHError(e, "failed to subscribe to the remote table with the provided ticket.") from e
+
+    def snapshot(self, ticket: bytes) -> Table:
+        """ Returns a snapshot of a published remote table with the given ticket.
+
+        Note, if the remote table is closed or its owner session is closed, the ticket becomes invalid. If the same
+        ticket is snapshot multiple times, multiple snapshots will be created.
+
+        Args:
+            ticket (bytes): the bytes of the ticket
+
+        Returns:
+            a Table
+
+        Raises:
+            DHError
+        """
+        try:
+            j_table_handle = self.j_session.of(_JTableSpec.ticket(ticket))
+            j_barrage_snapshot = self.j_barrage_session.snapshot(j_table_handle, _JBarrageTableResolver.SNAP_OPTIONS)
```

Comment:
Sorry I didn't notice this earlier; we'll be leaking exports this way. We should be calling snapshot w/ the TableSpec directly.